claripy/ast/base.py

```diff
@@ -211,64 +213,41 @@ def __new__(cls, op, args, add_variables=None, hash=None, **kwargs): #pylint:dis
             ))
 
         kwargs['annotations'] = annotations
+        return cls.__init_with_annotations__(
+            op,
+            a_args,
+            arg_max_depth + 1,
+            uneliminatable_annotations,
+            relocatable_annotations,
+            **kwargs
+        )
 
-        cache = cls._hash_cache
-        if hash is not None:
-            h = hash
-        elif op in {'BVS', 'BVV', 'BoolS', 'BoolV', 'FPS', 'FPV'} and not annotations:
-            if op == "FPV" and a_args[0] == 0.0 and math.copysign(1, a_args[0]) < 0:
-                # Python does not distinguish between +0.0 and -0.0 so we add sign to tuple to distinguish
-                h = (op, kwargs.get('length', None), ("-",) + a_args)
-            elif op == "FPV" and math.isnan(a_args[0]):
-                # cannot compare nans
-                h = (op, kwargs.get('length', None), ('nan',) + a_args[1:])
-            else:
-                h = (op, kwargs.get('length', None), a_args)
-
-            cache = cls._leaf_cache
-        else:
-            h = Base._calc_hash(op, a_args, kwargs) if hash is None else hash
+    @classmethod
+    def __init_with_annotations__(cls, op, a_args, depth=None, uneliminatable_annotations=None,
```

Comment:
`__new_with_annotations__`?

Reply:
That is a better name for this function, in my opinion. If we want to change that I'll need to see where else it is used and do a synced change